File: src/realtime/process.py
import numpy as np
from src.config import PreprocessConfig
from .buffers import CircularBuffer
from .dsp import compute_bandpowers

class StreamProcessor:
    def __init__(self, cfg: PreprocessConfig):
        self.cfg = cfg
        self.num_channels = len(cfg.channels)
        self.features_per_step = self.num_channels * 5 

        self.win_length = int(2.0 * cfg.fs)
        self.raw_history = CircularBuffer((self.num_channels, self.win_length))
        self.arousal_temporal = CircularBuffer((self.num_channels, cfg.win_samples))
        self.arousal_context = CircularBuffer((self.features_per_step, cfg.arousal_tensor_shape[1][2]))
        
        # Managing the identical 120-step doublet structure split across a 2D array matrix
        self.sleep_context = CircularBuffer((self.features_per_step * 2, 60))
        self.sleep_packet_counter = 0
        self.sample_counter = 0

    def push(self, sample: np.ndarray) -> None:
        self.raw_history.push(sample)
        self.arousal_temporal.push(sample)
        self.sample_counter += 1

        if self.sample_counter % self.cfg.hop_length == 0:
            self._compute_stft_step()

    def _compute_stft_step(self) -> None:
        history = self.raw_history.get_ordered()
        features = compute_bandpowers(history, self.cfg.fs, self.cfg.n_fft)
        self.arousal_context.push(features)
        
        # Replicate the Zig layout routing math
        t = self.sleep_packet_counter % 60
        offset = (self.sleep_packet_counter // 60) * 10
        
        self.sleep_context.buffer[offset:offset+10, t] = features
        self.sleep_packet_counter = (self.sleep_packet_counter + 1) % 120

    def extract_sleep_tensor(self) -> np.ndarray:
        """Unrolls the doublet context cleanly into Past (0:10) and Current (10:20)."""
        unrolled = np.zeros((20, 60), dtype=np.float32)
        
        for i in range(60):
            # Resolve the circular indices for the Oldest step (Past) and Newest step (Current)
            past_idx = (self.sleep_packet_counter + i) % 120
            curr_idx = (self.sleep_packet_counter + 60 + i) % 120
            
            # Map physical locations in the buffer
            past_row_base = (past_idx // 60) * 10
            past_col = past_idx % 60
            
            curr_row_base = (curr_idx // 60) * 10
            curr_col = curr_idx % 60
            
            # Pack cleanly into the top and bottom halves
            unrolled[0:10, i] = self.sleep_context.buffer[past_row_base : past_row_base + 10, past_col]
            unrolled[10:20, i] = self.sleep_context.buffer[curr_row_base : curr_row_base + 10, curr_col]
            
        return unrolled

File: deploy/src/process.zig
const std = @import("std");
const dsp = @import("dsp.zig");

pub const PipelineBuffers = struct {
    // Shared Temporal Data (100Hz)
    raw_history: [2][200]f32 = [_][200]f32{[_]f32{0} ** 200} ** 2,
    raw_idx: usize = 0,

    // Arousal Buffers
    arousal_temporal: [1][2][1500]f32 = undefined, // 15 seconds
    arousal_temporal_idx: usize = 0,
    arousal_context: [1][10][115]f32 = undefined, // ~1 min
    arousal_ctx_idx: usize = 0,

    // Sleep Buffers
    sleep_context: [1][20][60]f32 = undefined, // 20 features x 60 steps (100Hz, 50-sample hops)
    sleep_ctx_idx: usize = 0,

    stft_counter: usize = 0,

    pub fn push_sample(self: *@This(), ch0: f32, ch1: f32) void {
        self.raw_history[0][self.raw_idx] = ch0;
        self.raw_history[1][self.raw_idx] = ch1;

        self.arousal_temporal[0][0][self.arousal_temporal_idx] = ch0;
        self.arousal_temporal[0][1][self.arousal_temporal_idx] = ch1;
        self.arousal_temporal_idx = (self.arousal_temporal_idx + 1) % 1500;

        self.raw_idx = (self.raw_idx + 1) % 200;
        self.stft_counter += 1;

        if (self.stft_counter >= 50) {
            self.stft_counter = 0;
            self.compute_and_push_context(true);
            self.compute_and_push_context(false);
        }
    }

    fn compute_and_push_context(self: *@This(), is_arousal: bool) void {
        var ordered_ch0 = [_]f32{0} ** 200;
        var ordered_ch1 = [_]f32{0} ** 200;

        for (0..200) |i| {
            const idx = (self.raw_idx + i) % 200;
            ordered_ch0[i] = self.raw_history[0][idx];
            ordered_ch1[i] = self.raw_history[1][idx];
        }

        var bands0 = [_]f32{0} ** 5;
        var bands1 = [_]f32{0} ** 5;
        dsp.compute_bandpowers(&ordered_ch0, &bands0);
        dsp.compute_bandpowers(&ordered_ch1, &bands1);

        if (is_arousal) {
            for (0..5) |b| {
                self.arousal_context[0][b][self.arousal_ctx_idx] = bands0[b];
                self.arousal_context[0][b + 5][self.arousal_ctx_idx] = bands1[b];
            }
            self.arousal_ctx_idx = (self.arousal_ctx_idx + 1) % 115;
        } else {
            // Pack incoming features safely into the circular buffer layout
            const t = self.sleep_ctx_idx % 60;
            const feature_offset = (self.sleep_ctx_idx / 60) * 10;

            for (0..5) |b| {
                self.sleep_context[0][feature_offset + b][t] = bands0[b];
                self.sleep_context[0][feature_offset + 5 + b][t] = bands1[b];
            }
            self.sleep_ctx_idx = (self.sleep_ctx_idx + 1) % 120;
        }
    }

    pub fn prep_tensors_for_inference(self: *@This()) void {
        // --- 1. AROUSAL TEMPORAL ---
        var temp = self.arousal_temporal;
        for (0..1500) |i| {
            const src_idx = (self.arousal_temporal_idx + i) % 1500;
            self.arousal_temporal[0][0][i] = temp[0][0][src_idx];
            self.arousal_temporal[0][1][i] = temp[0][1][src_idx];
        }
        dsp.zscore_normalize(&self.arousal_temporal[0][0]);
        dsp.zscore_normalize(&self.arousal_temporal[0][1]);

        // --- 2. AROUSAL CONTEXT ---
        var ctx_temp = self.arousal_context;
        for (0..10) |c| {
            for (0..115) |i| {
                const src_idx = (self.arousal_ctx_idx + i) % 115;
                self.arousal_context[0][c][i] = ctx_temp[0][c][src_idx];
            }
            dsp.zscore_normalize(&self.arousal_context[0][c]);
        }

        // --- 3. SLEEP CONTEXT CHRONOLOGICAL UNROLLING ---
        var sleep_temp = self.sleep_context;

        for (0..60) |i| {
            // Resolve the circular indices for the Oldest step (Past) and Newest step (Current)
            const past_idx = (self.sleep_ctx_idx + i) % 120;
            const curr_idx = (self.sleep_ctx_idx + 60 + i) % 120;

            // Map physical locations in the buffer
            const past_row_base = (past_idx / 60) * 10;
            const past_col = past_idx % 60;

            const curr_row_base = (curr_idx / 60) * 10;
            const curr_col = curr_idx % 60;

            // Pack cleanly into the top and bottom halves
            for (0..10) |sub| {
                self.sleep_context[0][sub][i] = sleep_temp[0][past_row_base + sub][past_col];
                self.sleep_context[0][10 + sub][i] = sleep_temp[0][curr_row_base + sub][curr_col];
            }
        }

        // Apply Z-Score Normalization
        for (0..20) |f| {
            var feature_slice = [_]f32{0} ** 60;

            // Extract single feature timeline
            for (0..60) |t| {
                feature_slice[t] = self.sleep_context[0][f][t];
            }

            dsp.zscore_normalize(&feature_slice);

            // Repack normalized
            for (0..60) |t| {
                self.sleep_context[0][f][t] = feature_slice[t];
            }
        }
    }
};
